weave/weave_types.py

```diff
@@ -144,6 +147,16 @@ def type_from_dict(d: typing.Union[str, dict]) -> "Type":
             raise errors.WeaveSerializeError("Can't deserialize type from: %s" % d)
         return type_.from_dict(d)
 
+    @staticmethod
+    def type_from_dict_use_cache(d: typing.Union[str, dict]) -> "Type":
+        cache_key = id(d)
```

Comment:
I chose to use the `id()` of the dict as the cache key because `serialize2` ensures that objects with the same value share the same object reference.

However, if we want to use this in other places, we may want to do some value-based hash.

Reply:
`id` can be a foot gun in python as the id is allowed to be re-used after gcing the object.

Reply:
Good catch! Thanks for lmk.

Should be addressed via the context block you suggested.